vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -1,1011 +1,527 @@
+                tags.extend(self.base_tags)
 
-        # ## <TEST-INSTRUMENTATION>
-        custom_tags = instance.get('tags', []) + ['instance:{}'.format(i_key)]
-        self.histogram('datadog.agent.vsphere.metric_colection.time', t.total(), tags=custom_tags)
-        # ## </TEST-INSTRUMENTATION>
+                # vsphere "rates" should be submitted as gauges (rate is
+                # precomputed).
+                self.gauge(ensure_unicode(metric_name), value, hostname=hostname, tags=tags)
 
-    def collect_metrics(self, instance):
+    def query_metrics_wrapper(self, query_specs):
+        """Just an instrumentation wrapper around the VSphereAPI.query_metrics method
+        Warning: called in threads
         """
-        Calls asynchronously _collect_metrics_async on all MORs, as the
-        job queue is processed the Aggregator will receive the metrics.
+        t0 = Timer()
+        metrics_values = self.api.query_metrics(query_specs)
+        self.histogram('datadog.vsphere.query_metrics.time', t0.total(), tags=self.base_tags, raw=True)
+        return metrics_values
+
+    def collect_metrics(self, thread_pool):
+        """Creates a pool of threads and run the query_metrics calls in parallel."""
+        tasks = []
+        for resource_type in self.collected_resource_types:
+            mors = self.infrastructure_cache.get_mors(resource_type)
+            counters = self.metrics_metadata_cache.get_metadata(resource_type)
+            metric_ids = []
+            for counter_key, metric_name in iteritems(counters):
+                instance = ""
+                if should_collect_per_instance_values(metric_name, resource_type):
+                    instance = "*"
+                metric_ids.append(vim.PerformanceManager.MetricId(counterId=counter_key, instance=instance))
+
+            for batch in self.make_batch(mors, metric_ids, resource_type):
+                query_specs = []
+                for mor, metrics in iteritems(batch):
+                    query_spec = vim.PerformanceManager.QuerySpec()
+                    query_spec.entity = mor
+                    query_spec.metricId = metrics
+                    if resource_type in REALTIME_RESOURCES:
+                        query_spec.intervalId = REALTIME_METRICS_INTERVAL_ID
+                        query_spec.maxSample = 1  # Request a single datapoint
+                    else:
+                        # We cannot use `maxSample` for historical metrics, let's specify a timewindow that will
+                        # contain at least one element
+                        query_spec.startTime = datetime.now() - timedelta(hours=2)
+                    query_specs.append(query_spec)
+                if query_specs:
+                    tasks.append(thread_pool.submit(lambda q: self.query_metrics_wrapper(q), query_specs))
+
+        self.log.info("Queued all %d tasks, waiting for completion.", len(tasks))
+        while tasks:
```

Comment:
💯 Thanks!

Reply:
(minor) I would pass directly the `results` to `submit_metrics_callback` instead of the `future`. That will help isolate async code in `collect_metrics_async`.